### PR TITLE
[DOWNSTREAM] Cyrillic Support v2

### PR DESCRIPTION
**Pull request by larentoun**
## What Does This PR Do

Original: https://github.com/ParadiseSS13/Paradise/pull/21638

- Adds `_char` to most `copytext` and other things
- Adds `utf-8` to html

fixes https://github.com/ParadiseSS13/Paradise/issues/23731

## Why It's Good For The Game

Russian downstream need this to minimize direct changes to the code for their fork.

Makes it easier for downstream to pull changes, so it's more up-to-date with current Paradise and with less changes -> easier to maintain both the fork and the upstream

Also utf-8 support

## Testing

- [x] Cyrillic names for Prefs, Syndi-IDs
- [x] Can use chat freely on cyrillic
- [x] Renamed papers and wrote in Cyrillic
- [x] PDAs work
- [x] Can rename as silicon  
- [x] ViewVariables works for cyrillic text
- [x] Notes (memories) work for cyrillic text  
- [x] PlayerPanel works for cyrillic text
- [x] Status Displays work
- [x] Character Setup work for cyrillic text 
- [x] Character Records work for cyrillic text 
- [x] Silicon Laws displayed both in chat and in Law Manager for cyrillic text 
- [x] Spellbook supports cyrillic 

## Changelog
:cl:
add: Cyrillic Support for a lot of things, mostly names and capitalize
add: UTF-8 support for HTML
/:cl:

### CODE REVIEW

**Comment by larentoun** [comment]
@Gaxeer @Furrior @Legendaxe

File: code/game/objects/items/weapons/AI_modules.dm [hunk 1]
```diff
@@ -212,7 +212,7 @@ AI MODULES
 	if(new_lawpos < MIN_SUPPLIED_LAW_NUMBER)	return
 	lawpos = min(new_lawpos, MAX_SUPPLIED_LAW_NUMBER)
 	var/newlaw = ""
-	var/targName = sanitize(copytext(input(usr, "Please enter a new law for the AI.", "Freeform Law Entry", newlaw),1,MAX_MESSAGE_LEN))
+	var/targName = sanitize(copytext_char(input(usr, "Please enter a new law for the AI.", "Freeform Law Entry", newlaw),1,MAX_MESSAGE_LEN))
```
Comment (Burzah) [inline code/game/objects/items/weapons/AI_modules.dm:215]:
```suggestion
	var/targName = sanitize(copytext_char(input(usr, "Please enter a new law for the AI.", "Freeform Law Entry", newlaw), 1, MAX_MESSAGE_LEN))
```

File: code/game/objects/items/devices/paicard.dm [hunk 2]
```diff
@@ -281,7 +281,7 @@
 			if(2)
 				radio.ToggleReception()
 	if(href_list["setlaws"])
-		var/newlaws = sanitize(copytext(input("Enter any additional directives you would like your pAI personality to follow. Note that these directives will not override the personality's allegiance to its imprinted master. Conflicting directives will be ignored.", "pAI Directive Configuration", pai.pai_laws) as message,1,MAX_MESSAGE_LEN))
+		var/newlaws = sanitize(copytext_char(input("Enter any additional directives you would like your pAI personality to follow. Note that these directives will not override the personality's allegiance to its imprinted master. Conflicting directives will be ignored.", "pAI Directive Configuration", pai.pai_laws) as message,1,MAX_MESSAGE_LEN))
```
Comment (Burzah) [inline code/game/objects/items/devices/paicard.dm:284]:
```suggestion
		var/newlaws = sanitize(copytext_char(input("Enter any additional directives you would like your pAI personality to follow. Note that these directives will not override the personality's allegiance to its imprinted master. Conflicting directives will be ignored.", "pAI Directive Configuration", pai.pai_laws) as message, 1, MAX_MESSAGE_LEN))
```

File: code/game/machinery/camera/camera.dm [hunk 3]
```diff
@@ -180,10 +180,10 @@
 					to_chat(AI, "<b>[U]</b> holds <a href='?_src_=usr;show_paper=1;'>\a [itemname]</a> up to one of your cameras ...")
 				else
 					to_chat(AI, "<b><a href='?src=[AI.UID()];track=[html_encode(U.name)]'>[U]</a></b> holds <a href='?_src_=usr;show_paper=1;'>\a [itemname]</a> up to one of your cameras ...")
-				AI.last_paper_seen = "<HTML><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>"
+				AI.last_paper_seen = "<HTML><meta charset='utf-8'><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>"
 			else if(O.client && O.client.eye == src)
 				to_chat(O, "[U] holds \a [itemname] up to one of the cameras ...")
-				O << browse("<HTML><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>", "window=[itemname]")
+				O << browse("<HTML><meta charset='utf-8'><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>", "window=[itemname]")
```
Comment (Burzah) [inline code/game/machinery/camera/camera.dm:186]:
```suggestion
				O << browse("<html><meta charset='utf-8'><head><title>[itemname]</title></head><body><tt>[info]</tt></body></html>", "window=[itemname]")
```

File: code/game/machinery/camera/camera.dm [hunk 4]
```diff
@@ -180,10 +180,10 @@
 					to_chat(AI, "<b>[U]</b> holds <a href='?_src_=usr;show_paper=1;'>\a [itemname]</a> up to one of your cameras ...")
 				else
 					to_chat(AI, "<b><a href='?src=[AI.UID()];track=[html_encode(U.name)]'>[U]</a></b> holds <a href='?_src_=usr;show_paper=1;'>\a [itemname]</a> up to one of your cameras ...")
-				AI.last_paper_seen = "<HTML><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>"
+				AI.last_paper_seen = "<HTML><meta charset='utf-8'><HEAD><TITLE>[itemname]</TITLE></HEAD><BODY><TT>[info]</TT></BODY></HTML>"
```
Comment (Burzah) [inline code/game/machinery/camera/camera.dm:183]:
```suggestion
				AI.last_paper_seen = "<html><meta charset='utf-8'><head><title>[itemname]</title></head><body><tt>[info]</tt></body></html>"
```

File: code/game/jobs/job_exp.dm [hunk 5]
```diff
@@ -52,7 +52,7 @@ GLOBAL_LIST_INIT(role_playtime_requirements, list(
 	if(!check_rights(R_ADMIN|R_MOD|R_MENTOR))
 		return
 	var/list/msg = list()
-	msg  += "<html><head><title>Playtime Report</title></head><body>"
+	msg  += "<html><meta charset='utf-8'><head><title>Playtime Report</title></head><body>"
 	var/datum/job/theirjob
 	var/jtext
 	msg += "<TABLE border ='1'><TR><TH>Player</TH><TH>Job</TH><TH>Crew</TH>"
```
Comment (Burzah) [inline code/game/jobs/job_exp.dm:58]:
```suggestion
	msg += "<table border ='1'><tr><th>Player</th><th>Job</th><th>Crew</th>"
```

File: code/game/gamemodes/wizard/wizard.dm [hunk 6]
```diff
@@ -86,7 +86,7 @@
 	var/wizard_name_first = pick(GLOB.wizard_first)
 	var/wizard_name_second = pick(GLOB.wizard_second)
 	var/randomname = "[wizard_name_first] [wizard_name_second]"
-	var/newname = sanitize(copytext(input(wizard_mob, "You are the Space Wizard. Would you like to change your name to something else?", "Name change", randomname) as null|text,1,MAX_NAME_LEN))
+	var/newname = sanitize(copytext_char(input(wizard_mob, "You are the Space Wizard. Would you like to change your name to something else?", "Name change", randomname) as null|text,1,MAX_NAME_LEN))
```
Comment (Burzah) [inline code/game/gamemodes/wizard/wizard.dm:89]:
```suggestion
	var/newname = sanitize(copytext_char(input(wizard_mob, "You are the Space Wizard. Would you like to change your name to something else?", "Name change", randomname) as null|text, 1, MAX_NAME_LEN))
```

File: code/game/gamemodes/wizard/artefact.dm [hunk 7]
```diff
@@ -48,7 +48,7 @@
 		var/wizard_name_first = pick(GLOB.wizard_first)
 		var/wizard_name_second = pick(GLOB.wizard_second)
 		var/randomname = "[wizard_name_first] [wizard_name_second]"
-		var/newname = sanitize(copytext(input(M, "You are the wizard's apprentice. Would you like to change your name to something else?", "Name change", randomname) as null|text,1,MAX_NAME_LEN))
+		var/newname = sanitize(copytext_char(input(M, "You are the wizard's apprentice. Would you like to change your name to something else?", "Name change", randomname) as null|text,1,MAX_NAME_LEN))
```
Comment (Burzah) [inline code/game/gamemodes/wizard/artefact.dm:51]:
```suggestion
		var/newname = sanitize(copytext_char(input(M, "You are the wizard's apprentice. Would you like to change your name to something else?", "Name change", randomname) as null|text, 1, MAX_NAME_LEN))
```

File: code/game/gamemodes/nuclear/nuclear.dm [hunk 8]
```diff
@@ -424,7 +424,7 @@
 
 /proc/nukelastname(mob/M as mob) //--All praise goes to NEO|Phyte, all blame goes to DH, and it was Cindi-Kate's idea. Also praise Urist for copypasta ho.
 	var/randomname = pick(GLOB.last_names)
-	var/newname = sanitize(copytext(input(M,"You are the nuke operative [pick("Czar", "Boss", "Commander", "Chief", "Kingpin", "Director", "Overlord")]. Please choose a last name for your family.", "Name change",randomname),1,MAX_NAME_LEN))
+	var/newname = sanitize(copytext_char(input(M,"You are the nuke operative [pick("Czar", "Boss", "Commander", "Chief", "Kingpin", "Director", "Overlord")]. Please choose a last name for your family.", "Name change",randomname),1,MAX_NAME_LEN))
```
Comment (Burzah) [inline code/game/gamemodes/nuclear/nuclear.dm:427]:
```suggestion
	var/newname = sanitize(copytext_char(input(M,"You are the nuke operative [pick("Czar", "Boss", "Commander", "Chief", "Kingpin", "Director", "Overlord")]. Please choose a last name for your family.", "Name change", randomname), 1, MAX_NAME_LEN))
```

File: code/datums/mind.dm [hunk 9]
```diff
@@ -762,7 +762,7 @@
 				var/datum/objective/escape/escape_with_identity/O = new_objective
 				O.target_real_name = new_objective.target.current.real_name
 			if("custom")
-				var/expl = sanitize(copytext(input("Custom objective:", "Objective", objective ? objective.explanation_text : "") as text|null,1,MAX_MESSAGE_LEN))
+				var/expl = sanitize(copytext_char(input("Custom objective:", "Objective", objective ? objective.explanation_text : "") as text|null,1,MAX_MESSAGE_LEN))
```
Comment (Burzah) [inline code/datums/mind.dm:765]:
```suggestion
				var/expl = sanitize(copytext_char(input("Custom objective:", "Objective", objective ? objective.explanation_text : "") as text|null, 1, MAX_MESSAGE_LEN))
```

File: code/datums/mind.dm [hunk 10]
```diff
@@ -524,7 +524,7 @@
 		alert("Not before round-start!", "Alert")
 		return
 
-	var/list/out = list("<B>[name]</B>[(current && (current.real_name != name))?" (as [current.real_name])" : ""]")
+	var/list/out = list("<meta charset='UTF-8'><B>[name]</B>[(current && (current.real_name != name))?" (as [current.real_name])" : ""]")
```
Comment (Burzah) [inline code/datums/mind.dm:527]:
```suggestion
	var/list/out = list("<meta charset='UTF-8'><b>[name]</b>[(current && (current.real_name != name))?" (as [current.real_name])" : ""]")
```

File: code/datums/mind.dm [hunk 11]
```diff
@@ -185,7 +185,7 @@
 	if(!recipient)
 		recipient = current
 	var/list/output = list()
-	output.Add("<B>[current.real_name]'s Memories:</B><HR>")
+	output.Add("<meta charset='UTF-8'><B>[current.real_name]'s Memories:</B><HR>")
```
Comment (Burzah) [inline code/datums/mind.dm:188]:
```suggestion
	output.Add("<meta charset='UTF-8'><b>[current.real_name]'s Memories:</b><hr>")
```

File: code/__HELPERS/text.dm [hunk 12]
```diff
@@ -364,7 +364,7 @@
 		else
 			break
 	if(max_length)
-		input = copytext(input,1,max_length)
+		input = copytext_char(input,1,max_length)
```
Comment (Burzah) [inline code/__HELPERS/text.dm:367]:
```suggestion
		input = copytext_char(input, 1, max_length)
```

**Review by Burzah — CHANGES_REQUESTED** [review]
Okay I did a quick pass at this. Most of the changes I am requesting are for formatting and some code cleanliness changes like HTML tag formatting. Some of the stuff you did not touch yourself, but while we are here we can make it better. Thanks!

File: code/datums/mind.dm [hunk 13]
```diff
@@ -185,7 +185,7 @@
 	if(!recipient)
 		recipient = current
 	var/list/output = list()
-	output.Add("<B>[current.real_name]'s Memories:</B><HR>")
+	output.Add("<meta charset='UTF-8'><b>[current.real_name]'s Memories:</b><hr>")
```
Comment (S34NW) [inline code/datums/mind.dm:188]: Are all these charset specifications required? HTML5 defaults to `UTF-8` https://www.w3schools.com/charsets/ref_html_utf8.asp
Having said that, I know BYOND is often unexpected.

Comment (larentoun) [reply code/datums/mind.dm:188]: > Are all these charset specifications required? HTML5 defaults to `UTF-8` https://www.w3schools.com/charsets/ref_html_utf8.asp Having said that, I know BYOND is often unexpected.

Yes, excatly that - we need to specify for byond. You can easily check it by using any cyrillic in htmls

Example is https://github.com/ss220club/Paradise-SS220/pull/1014 , where specifying is actually fixing stuff

**Review by S34NW — APPROVED** [review]

**Comment by S34NW** [comment]
DNM until TM is conducted.

Voting process is bespoke for this PR, a full headcoder vote is underway.

**Review by S34NW — APPROVED** [review]
PR has demonstrated only minimal performance impact, HC vote passed. This is now good for merge.

**Pull request merged**